### PR TITLE
fix: unable to override timeout

### LINKED ISSUE #1020: Can't override timeout if program specified it; can't override urgency
### Issue description

I'm trying to override the timeout for some notifications. It does not work when the application specified a timeout. I also discovered that I can not override the urgency.

* `notify-send -a "program" "summary" "body"` — timeout override works, urgency override does not work.
* `notify-send -a "program" -t 10000 "summary" "body"` — timeout override does not work, urgency override does not work.

#### Example notification

This is the output I get with `notify-send -a "program" -t 10000 "summary" "body"` The urgency seems to be set correctly, but it is displayed as a normal notidication.

```
{
        appname: 'program'
        summary: 'summary'
        body: 'body'
        icon: 'dialog-warning'
        raw_icon set: false
        icon_id: '(null)'
        desktop_entry: ''
        category:
        timeout: 10000
        urgency: CRITICAL
        transient: 0
        formatted: '<b>summary</b>
body'
        fg: #ffffff
        bg: #00ff00
        highlight: #1745d1
        frame: #888888
        fullscreen: show
        format: <b>%s</b>
%b
        progress: -1
        stack_tag:
        id: 2
        actions: {}
        script_count: 0
}
```

### Installation info

- Version: `Dunst - A customizable and lightweight notification-daemon 1.7.3 (2021-12-08)`
- Install type: `Gentoo repo`
- Window manager / Desktop environment: `i3`

<details>
<summary>Minimal dunstrc</summary>

```ini
[urgency_normal]
    background  = "#00ff00"
    timeout = 10

[urgency_critical]
    background  = "#ff0000"
    timeout = 0

[override_test]
    appname = "program"
    timeout = 2
    urgency = critical
```
</details>

### PR DESCRIPTION
**Pull request by rpbranco**
Fixes: #1020

### CODE REVIEW

**Comment by tastytea** [comment]
Thanks! I can can confirm the timeout override works with this patch.

But when I override the urgency to be critical only the icon changes to the one used for critical notifications, the colors are the same as for normal ones.

**Comment by rpbranco** [comment]
No worries! I hope it was helpful :smile:. 

Regarding your problem with the notification colour, the [documentation](https://dunst-project.org/documentation/) states:

> `urgency`
> This sets the notification urgency.
> 
> IMPORTANT NOTE: This currently DOES NOT re-apply the attributes from the urgency_* sections. The changed urgency will only be visible in rules defined later. Use msg_urgency to match it.

This leads me to believe that this is not a bug but intended behaviour.
So, in your example, you can have the proper notification colours if you swap the order of the sections.

Example:
```
[override_test]
    appname = "program"
    timeout = 2
    urgency = critical

[urgency_normal]
    background  = "#00ff00"
    timeout = 10

[urgency_critical]
    background  = "#ff0000"
    timeout = 0
```

**Comment by tastytea** [comment]
> So, in your example, you can have the proper notification colours if you swap the order of the sections.

Ah right, that works. 😊

**Comment by fwsmit** [comment]
Thanks for the fix. I can confirm that it's indeed intended behaviour that the rule order matters.

File: test/dbus.c [hunk 1]
```diff
@@ -654,18 +654,30 @@ TEST test_timeout_overflow(void)
         n_dbus = dbus_notification_new();
         n_dbus->app_name = "dunstteststack";
         n_dbus->app_icon = "NONE";
-        n_dbus->summary = "test_hint_urgency";
+        n_dbus->summary = "test_timeout_overflow";
         n_dbus->body = "Summary of it";
         n_dbus->expire_timeout = 2147484;
         gint64 expected_timeout = G_GINT64_CONSTANT(2147484000);
 
+        // Disable urgency_normal rule in order to avoid timeout override
```
Comment (fwsmit) [inline test/dbus.c:662]: This shouldn't be needed right? The dbus timeout should overwrite the rule.

Comment (rpbranco) [reply test/dbus.c:662]: > This shouldn't be needed right?

I did not know that the dbus timeout should have priority over the ones defined in the rules.

The change I made sets the notification timeout before applying the rules.
This gives priority to the timeouts defined in rules and not the one in the dbus.
That is why I had to change the test.

> The dbus timeout should overwrite the rule.

This makes it impossible for users to change a dbus timeout right?

**Comment by fwsmit** [comment]
As I already said earlier, the current behaviour is intended. If there is anything you want to be changed, please discuss it in #1020. I'll close this MR for now.

**Pull request closed without merging**